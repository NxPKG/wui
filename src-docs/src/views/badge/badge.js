import React, { Fragment, useState } from 'react';

import {
  WuiBadge,
  WuiFlexItem,
  WuiFlexGroup,
  WuiSpacer,
  WuiSwitch,
  WuiText,
  WuiTitle,
} from '../../../../src/components';

const badges = [
  'default',
  'hollow',
  'primary',
  'secondary',
  'accent',
  'warning',
  'danger',
];

const customBadges = [
  '#DDD',
  '#AAA',
  '#666',
  '#333',
  '#BADA55',
  '#FCF7BC',
  '#FEA27F',
  '#FFA500',
  '#0000FF',
];

export default () => {
  const [isDisabled, setDisabled] = useState(false);

  return (
    <Fragment>
      <WuiTitle size="xs">
        <h2>Accepted color names</h2>
      </WuiTitle>
      <WuiSpacer size="m" />
      <WuiFlexGroup wrap responsive={false} gutterSize="xs">
        {badges.map(badge => (
          <WuiFlexItem grow={false} key={badge}>
            <WuiBadge color={badge}>{badge}</WuiBadge>
          </WuiFlexItem>
        ))}
      </WuiFlexGroup>
      <WuiSpacer />
      <WuiTitle size="xs">
        <h3>Custom color examples</h3>
      </WuiTitle>
      <WuiSpacer size="m" />
      <WuiFlexGroup
        wrap
        responsive={false}
        gutterSize="xs"
        style={{ maxWidth: '300px' }}>
        {customBadges.map(badge => (
          <WuiFlexItem grow={false} key={badge}>
            <WuiBadge color={badge}>{badge}</WuiBadge>
          </WuiFlexItem>
        ))}
      </WuiFlexGroup>
      <WuiSpacer />
      <WuiTitle size="xs">
        <h3>Disabled state</h3>
      </WuiTitle>
      <WuiSpacer size="m" />
      <WuiText size="s">
        Regardless of the assigned color, all badges use the same disabled state
        styles.
      </WuiText>
      <WuiSpacer size="m" />
      <WuiSwitch
        label="Show disabled state"
        checked={isDisabled}
        onChange={e => setDisabled(e.target.checked)}
      />
      <WuiSpacer size="m" />
      <WuiFlexGroup wrap responsive={false} gutterSize="xs">
        <WuiFlexItem grow={false}>
          <WuiBadge color="secondary" isDisabled={isDisabled}>
            {isDisabled ? 'Disabled badge' : 'Disable me!'}
          </WuiBadge>
        </WuiFlexItem>
      </WuiFlexGroup>
    </Fragment>
  );
};
