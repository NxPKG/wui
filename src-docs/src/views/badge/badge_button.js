import React from 'react';

import {
  WuiBadge,
  WuiFlexGroup,
  WuiFlexItem,
} from '../../../../src/components';

export default () => (
  <WuiFlexGroup wrap responsive={false} gutterSize="xs">
    <WuiFlexItem grow={false}>
      <WuiBadge
        color="primary"
        onClick={() => window.alert('Badge clicked')}
        onClickAriaLabel="Example of onClick event for the button"
        data-test-sub="testExample1">
        onClick on text within badge
      </WuiBadge>
    </WuiFlexItem>
    <WuiFlexItem grow={false}>
      <WuiBadge
        color="hollow"
        iconType="cross"
        iconSide="right"
        iconOnClick={() => window.alert('Icon inside badge clicked')}
        iconOnClickAriaLabel="Example of onClick event for icon within the button"
        data-test-sub="testExample2">
        onClick on icon within badge
      </WuiBadge>
    </WuiFlexItem>
    <WuiFlexItem grow={false}>
      <WuiBadge
        color="secondary"
        iconType="cross"
        iconSide="right"
        onClick={() => window.alert('Badge clicked')}
        onClickAriaLabel="Example of onClick event for the button"
        iconOnClick={() => window.alert('Icon inside badge clicked')}
        iconOnClickAriaLabel="Example of onClick event for icon within the button"
        data-test-sub="testExample3">
        onClick on both text and icon within badge
      </WuiBadge>
    </WuiFlexItem>
    <WuiFlexItem grow={false}>
      <WuiBadge
        isDisabled={true}
        color="danger"
        onClick={() => window.alert('Badge clicked')}
        onClickAriaLabel="Example of disabled button badge"
        iconOnClick={() => window.alert('Disabled badge clicked')}
        iconOnClickAriaLabel="Example of disabled button badge"
        data-test-sub="testExample4">
        disabled button badge
      </WuiBadge>
    </WuiFlexItem>
  </WuiFlexGroup>
);
