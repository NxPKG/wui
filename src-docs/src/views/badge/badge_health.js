import React from 'react';

import {
  WuiBadge,
  WuiFlexGroup,
  WuiFlexItem,
} from '../../../../src/components';

export default () => (
  <WuiFlexGroup wrap responsive={false} gutterSize="xs">
    <WuiFlexItem grow={false}>
      <WuiBadge color="secondary">Healthy</WuiBadge>
    </WuiFlexItem>
    <WuiFlexItem grow={false}>
      <WuiBadge color="warning">Warning</WuiBadge>
    </WuiFlexItem>
    <WuiFlexItem grow={false}>
      <WuiBadge color="danger">Critical</WuiBadge>
    </WuiFlexItem>
    <WuiFlexItem grow={false}>
      <WuiBadge color="default">Unknown</WuiBadge>
    </WuiFlexItem>
  </WuiFlexGroup>
);
