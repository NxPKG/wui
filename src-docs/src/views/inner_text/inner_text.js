import React, { useEffect, useState } from 'react';

import {
  WuiBadge,
  WuiCode,
  WuiHighlight,
  WuiHorizontalRule,
  WuiPanel,
  WuiText,
  WuiSpacer,
  WuiInnerText,
} from '../../../../src/components';

export default () => {
  const first = 'First';
  const second = 'Second';
  const [thing, setThing] = useState(first);
  const [[thing2, type], setThingAndType] = useState([first, 'span']);
  useEffect(() => {
    setTimeout(() => {
      const newThing = thing === second ? first : second;
      const newType = type === 'div' ? 'span' : 'div';
      setThing(newThing);
      setThingAndType([newThing, newType]);
    }, 5000);
  }, [thing, type]);

  return (
    <WuiText size="s">
      <p>
        <strong>Example:</strong>
      </p>
      <WuiInnerText>
        {(ref, innerText) => (
          <React.Fragment>
            <WuiPanel paddingSize="s" className="wui-displayInlineBlock">
              <span ref={ref} title={innerText}>
                Simple string content
              </span>
            </WuiPanel>
            <WuiSpacer />
            <p className="wui-displayInlineBlock">
              <strong>Output:</strong>
            </p>{' '}
            <WuiCode>{innerText}</WuiCode>
          </React.Fragment>
        )}
      </WuiInnerText>

      <WuiHorizontalRule margin="xl" />

      <p>
        <strong>Example with complex children:</strong>
      </p>
      <WuiInnerText>
        {(ref, innerText) => (
          <React.Fragment>
            <WuiPanel paddingSize="s" className="wui-displayInlineBlock">
              <span ref={ref} title={innerText}>
                <WuiHighlight search="content">
                  WuiHighlight content
                </WuiHighlight>{' '}
                <WuiBadge>with WuiBadge</WuiBadge>
              </span>
            </WuiPanel>
            <WuiSpacer />
            <p className="wui-displayInlineBlock">
              <strong>Output:</strong>
            </p>{' '}
            <WuiCode>{innerText}</WuiCode>
          </React.Fragment>
        )}
      </WuiInnerText>

      <WuiHorizontalRule margin="xl" />

      <p>
        <strong>Example with updating content:</strong>
      </p>
      <WuiInnerText>
        {(ref, innerText) => (
          <React.Fragment>
            <WuiPanel paddingSize="s" className="wui-displayInlineBlock">
              <span ref={ref} title={innerText}>
                {thing}
              </span>
            </WuiPanel>
            <WuiSpacer />
            <p className="wui-displayInlineBlock">
              <strong>Output:</strong>
            </p>{' '}
            <WuiCode>{innerText}</WuiCode>
          </React.Fragment>
        )}
      </WuiInnerText>

      <WuiHorizontalRule margin="xl" />

      <p>
        <strong>Example with updating element:</strong>
      </p>
      <WuiInnerText>
        {(ref, innerText) => (
          <React.Fragment>
            <WuiPanel paddingSize="s" className="wui-displayInlineBlock">
              {React.createElement(
                type,
                {
                  ref,
                  title: innerText,
                },
                thing2
              )}
            </WuiPanel>
            <WuiSpacer />
            <p className="wui-displayInlineBlock">
              <strong>Output:</strong>
            </p>{' '}
            <WuiCode>{innerText}</WuiCode>
          </React.Fragment>
        )}
      </WuiInnerText>
    </WuiText>
  );
};
