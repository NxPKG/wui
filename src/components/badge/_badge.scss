/**
 * 1. Accounts for the border
 */
.wuiBadge {
  font-size: $wuiFontSizeXS;
  font-weight: $wuiFontWeightMedium;
  line-height: $wuiSize + 2px; /* 1 */
  padding: 0 $wuiSizeS;
  display: inline-block;
  text-decoration: none;
  border-radius: $wuiBorderRadius / 2;
  border: solid 1px transparent;
  background-color: transparent;
  white-space: nowrap;
  vertical-align: middle;
  cursor: default;
  max-width: 100%;
  // The badge will only ever be as wide as its content
  // So, make the text left aligned to ensure all badges line up the same
  text-align: left;

  &.wuiBadge-isDisabled {
    // sass-lint:disable-block no-important
    // Using !important to override inline styles
    color: makeHighContrastColor($wuiButtonColorDisabled, $wuiButtonColorDisabled, 2) !important;
    background-color: $wuiButtonColorDisabled !important;
  }

  &:focus-within {
    @include wuiFocusRing('small');
  }

  + .wuiBadge {
    margin-left: $wuiSizeXS;
  }

  .wuiBadge__content {
    min-height: $wuiSize + ($wuiBorderWidthThin * 2); // Ensure proper height in case of just displaying an icon
    display: flex;
    align-items: center;
    overflow: hidden;
  }

  .wuiBadge__childButton {
    @include wuiTextTruncate;
    flex: 1 1 auto; // Must be 1 and 1 for IE to properly truncate
    text-align: inherit;
    font-weight: inherit;
    line-height: inherit;
    color: inherit;

    &:disabled {
      cursor: not-allowed;
    }

    &:not(:disabled) {
      &:hover,
      &:focus {
        text-decoration: underline;
      }
    }
  }

  .wuiBadge__iconButton {
    flex: 0 0 auto;
    font-size: 0; // Makes the button only as large as the icon so it aligns vertically better
    margin-left: $wuiSizeXS;

    &:focus {
      background-color: transparentize($wuiColorGhost, .2);
      color: $wuiColorInk;
      border-radius: 2px;
    }

    &:disabled {
      cursor: not-allowed;
    }

    .wuiBadge__icon {
      // Remove margins from icon itself so that focus state doesn't include that space
      margin: 0 !important; // sass-lint:disable-line no-important
    }
  }

  .wuiBadge__text {
    @include wuiTextTruncate;
    flex: 1 1 auto;
    cursor: default;
  }

  .wuiBadge__icon {
    flex: 0 0 auto;

    &:not(:only-child) {
      margin-left: $wuiSizeXS;
    }
  }

  &.wuiBadge--iconLeft {
    .wuiBadge__content {
      flex-direction: row-reverse;
    }

    .wuiBadge__iconButton,
    .wuiBadge__icon:not(:only-child) {
      margin-right: $wuiSizeXS;
      margin-left: 0;
    }
  }
}

.wuiBadge-isClickable {
  &:not(:disabled) {
    &:hover,
    &:focus {
      text-decoration: underline;
    }
  }

  &.wuiBadge-isDisabled {
    cursor: not-allowed;
  }

  &:focus {
    @include wuiFocusRing('small');
  }

  .wuiBadge__text {
    cursor: inherit;
  }
}

// Hollow has a border and is mostly used for autocompleters.
.wuiBadge--hollow {
  background-color: $wuiColorEmptyShade;
  border-color: lightOrDarkTheme($wuiBorderColor, tint($wuiBorderColor, 15%));
  color: $wuiTextColor;
}
