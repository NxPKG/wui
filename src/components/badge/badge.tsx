/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  AriaAttributes,
  FunctionComponent,
  HTMLAttributes,
  MouseEventHandler,
  ReactNode,
  Ref,
} from 'react';
import classNames from 'classnames';
import chroma from 'chroma-js';
import { CommonProps, ExclusiveUnion, keysOf, PropsOf } from '../common';
import {
  wuiPaletteColorBlindBehindText,
  getSecureRelForTarget,
  isColorDark,
} from '../../services';
import { WuiInnerText } from '../inner_text';
import { WuiIcon, IconColor, IconType } from '../icon';
import { chromaValid, parseColor } from '../color_picker/utils';
import { validateHref } from '../../services/security/href_validator';

type IconSide = 'left' | 'right';

type WithButtonProps = {
  /**
   * Will apply an onclick to the badge itself
   */
  onClick: MouseEventHandler<HTMLButtonElement>;

  /**
   * Aria label applied to the onClick button
   */
  onClickAriaLabel: AriaAttributes['aria-label'];
} & Omit<HTMLAttributes<HTMLButtonElement>, 'onClick' | 'color'>;

type WithAnchorProps = {
  href: string;
  target?: string;
  rel?: string;
} & Omit<HTMLAttributes<HTMLAnchorElement>, 'href' | 'color' | 'onClick'>;

type WithSpanProps = Omit<HTMLAttributes<HTMLSpanElement>, 'onClick' | 'color'>;

interface WithIconOnClick {
  /**
   * Will apply an onclick to icon within the badge
   */
  iconOnClick: MouseEventHandler<HTMLButtonElement>;

  /**
   * Aria label applied to the iconOnClick button
   */
  iconOnClickAriaLabel: AriaAttributes['aria-label'];
}

export type WuiBadgeProps = {
  /**
   * Accepts any string from our icon library
   */
  iconType?: IconType;

  /**
   * The side of the badge the icon should sit
   */
  iconSide?: IconSide;

  /**
   * Accepts either our palette colors (primary, secondary ..etc) or a hex value `#FFFFFF`, `#000`.
   */
  color?: IconColor;
  /**
   * Will override any color passed through the `color` prop.
   */
  isDisabled?: boolean;

  /**
   * Props passed to the close button.
   */
  closeButtonProps?: Partial<PropsOf<WuiIcon>>;
} & CommonProps &
  ExclusiveUnion<WithIconOnClick, {}> &
  ExclusiveUnion<
    ExclusiveUnion<WithButtonProps, WithAnchorProps>,
    WithSpanProps
  >;

// TODO - replace with variables once https://github.com/wazuh/wui/issues/2731 is closed
const colorInk = '#000';
const colorGhost = '#fff';

// The color blind palette has some stricter accessibility needs with regards to
// charts and contrast. We use the wuiPaletteColorBlindBehindText variant here since our
// accessibility concerns pertain to foreground (text) and background contrast
const visColors = wuiPaletteColorBlindBehindText();

const colorToHexMap: { [color in IconColor]: string } = {
  // TODO - replace with variable once https://github.com/wazuh/wui/issues/2731 is closed
  default: '#d3dae6',
  primary: visColors[1],
  secondary: visColors[0],
  accent: visColors[2],
  warning: visColors[5],
  danger: visColors[9],
};

export const COLORS = keysOf(colorToHexMap);

const iconSideToClassNameMap: { [side in IconSide]: string } = {
  left: 'wuiBadge--iconLeft',
  right: 'wuiBadge--iconRight',
};

export const ICON_SIDES = keysOf(iconSideToClassNameMap);

export const WuiBadge: FunctionComponent<WuiBadgeProps> = ({
  children,
  color = 'default',
  iconType,
  iconSide = 'left',
  className,
  isDisabled: _isDisabled,
  onClick,
  iconOnClick,
  onClickAriaLabel,
  iconOnClickAriaLabel,
  closeButtonProps,
  href,
  rel,
  target,
  style,
  ...rest
}) => {
  checkValidColor(color);

  const isHrefValid = !href || validateHref(href);
  const isDisabled = _isDisabled || !isHrefValid;

  let optionalCustomStyles: object | undefined = style;
  let textColor = null;
  // TODO - replace with variable once https://github.com/wazuh/wui/issues/2731 is closed
  const wcagContrastBase = 4.5; // WCAG AA contrast level
  let wcagContrast = null;
  let colorHex = null;

  // Check if a valid color name was provided
  if (COLORS.indexOf(color) > -1) {
    // Get the hex equivalent for the provided color name
    colorHex = colorToHexMap[color];

    // Set dark or light text color based upon best contrast
    textColor = setTextColor(colorHex);

    optionalCustomStyles = {
      backgroundColor: colorHex,
      color: textColor,
      ...optionalCustomStyles,
    };
  } else if (color !== 'hollow') {
    // This is a custom color that is neither from the base palette nor hollow
    // Let's do our best to ensure that it provides sufficient contrast

    // Set dark or light text color based upon best contrast
    textColor = setTextColor(color);

    // Check the contrast
    wcagContrast = getColorContrast(textColor, color);

    if (wcagContrast < wcagContrastBase) {
      // It's low contrast, so lets show a warning in the console
      console.warn(
        'Warning: ',
        color,
        ' badge has low contrast of ',
        wcagContrast.toFixed(2),
        '. Should be above ',
        wcagContrastBase,
        '.'
      );
    }

    optionalCustomStyles = {
      backgroundColor: color,
      color: textColor,
      ...optionalCustomStyles,
    };
  }

  const classes = classNames(
    'wuiBadge',
    {
      'wuiBadge-isClickable': (onClick || href) && !iconOnClick,
      'wuiBadge-isDisabled': isDisabled,
      'wuiBadge--hollow': color === 'hollow',
    },
    iconSideToClassNameMap[iconSide],
    className
  );

  const closeClassNames = classNames(
    'wuiBadge__icon',
    closeButtonProps && closeButtonProps.className
  );

  const Element = href && !isDisabled ? 'a' : 'button';
  const relObj: {
    href?: string;
    target?: string;
    rel?: string;
    onClick?:
      | ((event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void)
      | ((event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void);
  } = {};

  if (href && !isDisabled) {
    relObj.href = href;
    relObj.target = target;
    relObj.rel = getSecureRelForTarget({ href, target, rel });
  }
  if (onClick) {
    relObj.onClick = onClick;
  }

  let optionalIcon: ReactNode = null;
  if (iconType) {
    if (iconOnClick) {
      if (!iconOnClickAriaLabel) {
        console.warn(
          'When passing the iconOnClick props to WuiBadge, you must also provide iconOnClickAriaLabel'
        );
      }
      optionalIcon = (
        <button
          className="wuiBadge__iconButton"
          aria-label={iconOnClickAriaLabel}
          disabled={isDisabled}
          title={iconOnClickAriaLabel}
          onClick={iconOnClick}>
          <WuiIcon
            type={iconType}
            size="s"
            {...closeButtonProps}
            className={closeClassNames}
          />
        </button>
      );
    } else {
      optionalIcon = (
        <WuiIcon
          type={iconType}
          size={children ? 's' : 'm'}
          className="wuiBadge__icon"
        />
      );
    }
  }

  if (onClick && !onClickAriaLabel) {
    console.warn(
      'When passing onClick to WuiBadge, you must also provide onClickAriaLabel'
    );
  }

  const content = (
    <span className="wuiBadge__content">
      {children && <span className="wuiBadge__text">{children}</span>}
      {optionalIcon}
    </span>
  );

  if (iconOnClick) {
    return onClick || href ? (
      <span className={classes} style={optionalCustomStyles}>
        <span className="wuiBadge__content">
          <WuiInnerText>
            {(ref, innerText) => (
              <Element
                className="wuiBadge__childButton"
                disabled={isDisabled}
                aria-label={onClickAriaLabel}
                ref={ref}
                title={innerText}
                {...(relObj as HTMLAttributes<HTMLElement>)}
                {...(rest as HTMLAttributes<HTMLElement>)}>
                {children}
              </Element>
            )}
          </WuiInnerText>
          {optionalIcon}
        </span>
      </span>
    ) : (
      <WuiInnerText>
        {(ref, innerText) => (
          <span
            className={classes}
            style={optionalCustomStyles}
            ref={ref}
            title={innerText}
            {...rest}>
            {content}
          </span>
        )}
      </WuiInnerText>
    );
  } else if (onClick || href) {
    return (
      <WuiInnerText>
        {(ref, innerText) => (
          <Element
            disabled={isDisabled}
            aria-label={onClickAriaLabel}
            className={classes}
            style={optionalCustomStyles}
            ref={ref as Ref<HTMLButtonElement & HTMLAnchorElement>}
            title={innerText}
            {...(relObj as HTMLAttributes<HTMLElement>)}
            {...(rest as HTMLAttributes<HTMLElement>)}>
            {content}
          </Element>
        )}
      </WuiInnerText>
    );
  } else {
    return (
      <WuiInnerText>
        {(ref, innerText) => (
          <span
            className={classes}
            style={optionalCustomStyles}
            ref={ref}
            title={innerText}
            {...rest}>
            {content}
          </span>
        )}
      </WuiInnerText>
    );
  }
};

function getColorContrast(textColor: string, color: string) {
  const contrastValue = chroma.contrast(textColor, color);
  return contrastValue;
}

function setTextColor(bgColor: string) {
  const textColor = isColorDark(...chroma(bgColor).rgb())
    ? colorGhost
    : colorInk;

  return textColor;
}

function checkValidColor(color: null | IconColor | string) {
  const colorExists = !!color;
  const isNamedColor = (color && COLORS.includes(color)) || color === 'hollow';
  const isValidColorString = color && chromaValid(parseColor(color) || '');

  if (!colorExists && !isNamedColor && !isValidColorString) {
    console.warn(
      'WuiBadge expects a valid color. This can either be a three or six ' +
        `character hex value, rgb(a) value, hsv value, hollow, or one of the following: ${COLORS}. ` +
        `Instead got ${color}.`
    );
  }
}
