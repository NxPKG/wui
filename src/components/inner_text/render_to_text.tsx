/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { ReactNode, useCallback, useEffect, useRef } from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { useInnerText } from './inner_text';

export function useRenderToText(node: ReactNode, placeholder = ''): string {
  const [ref, text] = useInnerText(placeholder);
  const hostNode = useRef<Element | null>(null);

  const onUnmount = () => {
    if (hostNode.current) {
      unmountComponentAtNode(hostNode.current);
      hostNode.current = null;
    }
  };

  const setRef = useCallback(
    (node: Element | null) => {
      if (hostNode.current) {
        ref(node);
      }
    },
    [ref]
  );

  useEffect(() => {
    hostNode.current = (document.createDocumentFragment() as unknown) as Element;
    render(<div ref={setRef}>{node}</div>, hostNode.current);
    return () => {
      onUnmount();
    };
  }, [node, setRef]);

  return text || placeholder;
}
