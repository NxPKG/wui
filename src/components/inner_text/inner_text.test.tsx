/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { useState, useEffect } from 'react';
import { act } from 'react-dom/test-utils';
import { render, mount } from 'enzyme';
import { findTestSubject, requiredProps, sleep } from '../../test';

import { useInnerText, WuiInnerText } from './inner_text';
import { WuiBadge } from '../badge';

describe('useInnerText', () => {
  test('provides a callback `ref`', () => {
    let ref;
    const App = () => {
      [ref] = useInnerText();
      return <span />;
    };
    mount(<App />);

    expect(ref).toBeInstanceOf(Function);
  });

  test('provides the result of `innerText`', () => {
    const text = 'Test';
    let ref;
    let innerText;
    const App = () => {
      [ref, innerText] = useInnerText();
      return <span ref={ref}>{text}</span>;
    };
    mount(<App />);

    expect(innerText).toEqual(text);
  });

  test('accepts a fallback value', () => {
    const text = 'Test';
    const fallback = 'Fallback';
    let innerText;
    const App = () => {
      [, innerText] = useInnerText(fallback);
      return <span>{text}</span>;
    };
    mount(<App />);

    expect(innerText).toEqual(fallback);
  });

  test('handles updated elements', async () => {
    const timeout = 500;
    const first = 'First';
    const second = 'Second';
    let innerText;
    let ref;
    const App = () => {
      const [[thing, type], setThing] = useState([first, 'span']);
      useEffect(() => {
        setTimeout(() => {
          act(() => setThing([second, 'div']));
        }, timeout);
      }, [setThing]);
      [ref, innerText] = useInnerText();
      return (
        <div>
          {React.createElement(
            type,
            {
              ref,
              title: innerText,
            },
            thing
          )}
        </div>
      );
    };
    mount(<App />);

    expect(innerText).toEqual(first);

    await sleep(timeout + 10);

    expect(innerText).toEqual(second);
  });

  test('handles updated content', async () => {
    const timeout = 500;
    const first = 'First';
    const second = 'Second';
    let innerText;
    let ref;
    const App = () => {
      const [thing, setThing] = useState(first);
      useEffect(() => {
        setTimeout(() => {
          act(() => setThing(second));
        }, timeout);
      }, [setThing]);
      [ref, innerText] = useInnerText();
      return (
        <div>
          <span ref={ref} title={innerText}>
            {thing}
          </span>
        </div>
      );
    };
    mount(<App />);

    expect(innerText).toEqual(first);

    // MutationObserver polyfill institutes a 30ms mutation timeout period
    const mutationObserverPolyfillPeriod = 30;
    await sleep(timeout + mutationObserverPolyfillPeriod + 10);

    expect(innerText).toEqual(second);
  });
});

describe('WuiInnerText', () => {
  test('is rendered', () => {
    const component = render(
      <WuiInnerText {...requiredProps}>
        {(ref, innerText) => (
          <span ref={ref} title={innerText}>
            Test
          </span>
        )}
      </WuiInnerText>
    );

    expect(component).toMatchSnapshot();
  });

  test('uses innerText', () => {
    const text = 'Test';
    const component = mount(
      <WuiInnerText {...requiredProps}>
        {(ref, innerText) => (
          <span ref={ref} title={innerText} data-test-subj="span">
            {text}
          </span>
        )}
      </WuiInnerText>
    );

    const span = findTestSubject(component, 'span');
    expect(span.props().title).toBe(text);
  });

  test('accepts fallback prop', () => {
    const text = 'Test';
    const fallback = 'Fallback';
    const component = mount(
      <WuiInnerText {...requiredProps} fallback={fallback}>
        {(_, innerText) => (
          <span title={innerText} data-test-subj="span">
            {text}
          </span>
        )}
      </WuiInnerText>
    );

    const span = findTestSubject(component, 'span');
    expect(span.props().title).toBe(fallback);
  });

  test('works with wrapper and interspersed DOM elements', () => {
    const component = mount(
      <WuiInnerText {...requiredProps}>
        {(ref, innerText) => (
          <span ref={ref} title={innerText} data-test-subj="span">
            <div>
              I{' '}
              <span>
                can{' '}
                <em>
                  still <strong>read </strong>
                  <WuiBadge>this</WuiBadge>
                </em>
              </span>
            </div>
          </span>
        )}
      </WuiInnerText>
    );

    const span = findTestSubject(component, 'span');
    expect(span.props().title).toBe('I can still read this');
  });
});
