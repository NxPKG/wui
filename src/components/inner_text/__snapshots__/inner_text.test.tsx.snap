// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WuiInnerText is rendered 1`] = `
<span>
  Test
</span>
`;
